libs/rays/src/PathTracer.cpp

```diff
@@ -0,0 +1,480 @@
+                // necessarily normalized.
+                float3 n = normalize(float3 { rayhit.hit.Ng_x, rayhit.hit.Ng_y, rayhit.hit.Ng_z });
+                float3 b = randomPerp(n);
+                float3 t = cross(n, b);
+                mat3 tangentFrame = {t, b, n};
+
+                RTCRay aoray {
+                    .org_x = rayhit.ray.org_x + rayhit.ray.dir_x * rayhit.ray.tfar,
+                    .org_y = rayhit.ray.org_y + rayhit.ray.dir_y * rayhit.ray.tfar,
+                    .org_z = rayhit.ray.org_z + rayhit.ray.dir_z * rayhit.ray.tfar
+                };
+
+                float sum = 0;
+                for (size_t nsamp = 0, len = context->config.samplesPerPixel; nsamp < len; nsamp++) {
+                    const double2 u = hammersley(nsamp, inverseSampleCount);
+                    const float3 dir = tangentFrame * hemisphereCosSample(u);
+                    aoray.dir_x = dir.x;
+                    aoray.dir_y = dir.y;
+                    aoray.dir_z = dir.z;
+                    aoray.tnear = tnear;
+                    aoray.tfar = tfar;
+                    rtcOccluded1(embreeScene, &intersector, &aoray);
+                    if (aoray.tfar == -inf) {
+                        sum += 1.0f;
+                    }
+                }
+                image.getPixelRef(col, row)[0] = 1.0f - sum * inverseSampleCount;
+            }
+        }
+    }
+}
+
+static void renderTileToGbuffer(EmbreeContext* context, PixelRectangle rect) {
+    image::LinearImage& image = context->config.renderTarget;
+    image::LinearImage& gbuffer = context->gbuffer;
+    RTCScene embreeScene = context->scene;
+
+    const float tnear = context->config.aoRayNear;
+    const float tfar = context->config.aoRayFar;
+    const float iw = 1.0f / image.getWidth();
+    const float ih = 1.0f / image.getHeight();
+
+    auto generateOrthoCameraRay = [=] (uint16_t row, uint16_t col) {
+        return RTCRay {
+            .org_x = float(col) * iw,
+            .org_y = float(row) * ih,
+            .org_z = 1,
+            .tnear = 0,
+            .dir_x = 0,
+            .dir_y = 0,
+            .dir_z = -1,
+            .time = 0,
+            .tfar = inf,
+            .mask = 0xffffffff
+        };
+    };
+
+    for (size_t row = rect.topLeft.y, len = rect.bottomRight.y; row < len; ++row) {
+        for (size_t col = rect.topLeft.x, len = rect.bottomRight.x; col < len; ++col) {
+            RTCIntersectContext intersector;
+            rtcInitIntersectContext(&intersector);
+
+            RTCRay ray = generateOrthoCameraRay(row, col);
+            RTCRayHit rayhit { .ray = ray };
+            intersector.flags = RTC_INTERSECT_CONTEXT_FLAG_COHERENT;
+            rtcIntersect1(embreeScene, &intersector, &rayhit);
+
+            if (rayhit.ray.tfar != inf) {
+                float* position = gbuffer.getPixelRef(col, row);
+                float* normal = position + 3;
```

Comment:
Might be good to define the `3` here as a constant, which will help with readability and if the layout of the gbuffer ever changes.

Reply:
Done.